[IMP] spreadsheet: expanding pivot table documentation (#17973)
odoo/documentation: content/applications/productivity/spreadsheet/work_with_data/pivot_tables.rst

============
Pivot tables
============

> Pivot tables allow you to organize, summarize and analyze large amounts of data without the need for

Reviewer: ```suggestion
Pivot tables allow you to organize, summarize, and analyze large amounts of data without the need for
```

total amount or quantity ordered), the corresponding values of each intersecting data point are
calculated automatically.

Various manipulations can be performed to see the same dataset from different perspectives, such as
:ref:`rearranging or sorting dimensions <spreadsheet/pivot-tables/dimensions>`, or
:ref:`changing how measures are aggregated or calculated <spreadsheet/pivot-tables/measures>`.

.. example::
   In the example, the pivot table shows the :guilabel:`Total` amount of sales orders per
   salesperson per quarter. The :guilabel:`Salesperson` field is represented in rows, the
   :guilabel:`Order Date` (grouped by quarter and year) in columns, while the values represent the
   total amount for the given salesperson and quarter.

   .. image:: pivot_tables/pivot-table.png
      :alt: Pivot table showing sales per salesperson per quarter

With Odoo Spreadsheet, it is possible to:

- :ref:`insert a pivot table from an Odoo pivot view <spreadsheet/insert_pivot_table/from_view>`
  into an Odoo spreadsheet, Odoo dashboard, or quote calculator spreadsheet.

- :ref:`insert a pivot table referencing Odoo data directly from an Odoo spreadsheet

>   <spreadsheet/pivot-tables/create>`. This option allows you to retrieve and work with data from

Reviewer: ```suggestion
  <spreadsheet/insert_pivot_table/from_spreadsheet>`. This option allows you to retrieve and work with data from
```
I assume that's the ref you wanted :)

Author: Thanks!

  any model, including models for which no Odoo pivot view is available e.g., the *Sales Order Line*
  model.

- :ref:`insert a pivot table from a range of data <spreadsheet/pivot-tables/create>`. The data
  can be static data or dynamic Odoo data that had been inserted into the spreadsheet, e.g., from a
  list view.

When a pivot table is inserted in an Odoo spreadsheet, a :ref:`data source
<spreadsheet/insert/data-sources>` is created, which connects the pivot table to the underlying
database data or spreadsheet range.

.. seealso::
   :ref:`Static vs dynamic pivot tables <spreadsheet/insert-pivot-table/static-vs-dynamic>`

.. _spreadsheet/pivot-tables/create:

Insert a pivot table from a range of data
=========================================

.. tip::
   Before inserting a pivot table from a range of data:

>    - organize your data in columns, not rows

Reviewer: ```suggestion
   - Organize your data in columns, not rows.
```
I understand what you mean, but maybe spelling it out would be better for those who aren't familiar with pivot tables, adding something like: "_; each column should contain one type of information, while each row should contain one record._"

>    - ensure all columns have a unique header; avoid double rows of headers or merged cells

Reviewer: ```suggestion
   - Ensure all columns have a unique header; avoid double rows of headers or merged cells.
```

   - format your data as a table by selecting any cell in the range, clicking :icon:`os-edit-table`
     :guilabel:`Insert table` from the toolbar, then selecting a style. This ensures any updates to

>      the table are reflected in the pivot table

Reviewer: ```suggestion
   - Format your data as a table by selecting any cell in the range, clicking the :icon:`os-edit-table`
     :guilabel:`(Insert table)` icon from the toolbar, then selecting a style. This ensures any updates to
     the table are reflected in the pivot table.
```
Maybe add the _fa-caret-down_ icon as well? But then you technically click two icons 😅

Author: Like I said, LOVE icons so it's done!

To create a pivot table from a range of data:

#. Open the relevant Odoo spreadsheet.
#. Select any cell within the range of data to be analyzed, then click :menuselection:`Insert -->`
   :icon:`oi-view-pivot` :menuselection:`Pivot table` :icon:`fa-caret-right` :menuselection:`From
   range` from the menu bar.

> A new sheet opens, with an empty pivot table in the top-left corner of the sheet; the sheet name is

Reviewer: ```suggestion
A new sheet opens with an empty pivot table in the top-left corner of the sheet; the sheet name is
```

  if needed.
- :guilabel:`Model`: for a pivot table referencing Odoo data that was inserted from an Odoo pivot
  view or inserted directly from the spreadsheet, the model from which the data is retrieved.

.. _spreadsheet/pivot-tables/properties-domain:

- :guilabel:`Domain`: for a pivot table referencing Odoo data that was inserted from an Odoo pivot
  view or inserted directly from the spreadsheet, the rules used to determine which Odoo records
  are shown. Click :guilabel:`Edit domain` to add or edit rules.

  .. note::
     When referencing dynamic Odoo data in a pivot table and using :doc:`global filters
     <global_filters>`, this domain is combined with the selected values of the global filter before
     the data is loaded into the spreadsheet.

The :guilabel:`Columns` and :guilabel:`Rows` are the :ref:`dimensions
<spreadsheet/pivot-tables/dimensions>` used to group or segment data, while the :ref:`Measures
<spreadsheet/pivot-tables/measures>` define what is being measured, or analyzed, based on the
selected dimensions.

To :ref:`prevent real-time updates <spreadsheet/pivot-tables/properties-defer-updates>` while

> building or manipulating a pivot table, enable :guilabel:`Defer updates`.

Reviewer: I'm not sure it's worth keeping here since you already have the tip about deferring updates.

Author: I prefer to keep it here since this is an overview of what you see in the properties panel, so it would feel odd to not mention it.

To duplicate or delete *the data source* of a pivot table, click the :icon:`fa-cog`
:guilabel:`(gear)` icon at the top of the properties panel, then :icon:`fa-clone`
:guilabel:`Duplicate` or :icon:`fa-trash` :guilabel:`Delete` as relevant.

.. seealso::
   :ref:`Duplicate a pivot table <spreadsheet/pivot-tables/duplicate>`

>    :ref:`Delete a pivot table <spreadsheet/pivot-tables/delete>`

Reviewer: ```suggestion
   - :ref:`Duplicate a pivot table <spreadsheet/pivot-tables/duplicate>`
   - :ref:`Delete a pivot table <spreadsheet/pivot-tables/delete>`
```
Or you can insert a blank line; otherwise, the content will be displayed like this:
<img width="385" height="105" alt="Image" src="https://github.com/user-attachments/assets/603d0dd0-e642-45c1-ba6f-8d2c0a4a5538" />

Author: Ooops, had forgotten to put bullets! (I prefer that over the blank line)

level of granularity of the data.

To add a dimension to a pivot table:

#. Open the :ref:`pivot table's properties panel <spreadsheet/pivot-tables/properties>`.
#. In the :guilabel:`Columns` or :guilabel:`Rows` section, as relevant, click :guilabel:`Add`.
#. Select the appropriate field.
#. Change the :ref:`sorting <spreadsheet/pivot-tables/dimensions-sorting>`, if desired.

.. tip::
   - For date- or time-based dimensions, select the desired :guilabel:`Granularity` from the options
     in the dropdown menu.
   - In a pivot table referencing Odoo data that was inserted from an Odoo pivot view or inserted
     directly from the spreadsheet, click the :icon:`oi-chevron-right` :guilabel:`(right arrow)`
     next to the field name to access the list of related fields when adding columns or rows.

To remove a dimension from a pivot table, click the :icon:`fa-trash` :guilabel:`(delete)` icon on
the dimension's card.

.. _spreadsheet/pivot-tables/dimensions-add-nested:

Nested dimensions
*****************

Adding multiple row or columns dimensions creates a nested hierarchy. To change the hierarchy

> of the dimensions, drag the dimension's card to the desired position within its.

Reviewer: ```suggestion
Adding multiple row or column dimensions creates a nested hierarchy. To change the hierarchy
of the dimensions, drag the dimension's card to the desired position within its section.
```

:guilabel:`(expand)` or :icon:`fa-minus-square-o` :guilabel:`(collapse)` icon as relevant.

.. _spreadsheet/pivot-tables/dimensions-rearrange:

Rearrange dimensions
~~~~~~~~~~~~~~~~~~~~

Rearranging the dimensions of a pivot table, i.e., moving a field from a column to a row, or vice
versa, allows you to view the same dataset from different perspectives and potentially gain new
insights.

To change the axis on which an individual dimension is shown:

#. Open the :ref:`pivot table's properties panel <spreadsheet/pivot-tables/properties>`.
#. Drag the dimension's card from the :guilabel:`Columns` section to the :guilabel:`Rows` section or
   vice versa.

To move *all* the dimensions represented in columns to rows and vice versa, at the same time:

#. Open the :ref:`pivot table's properties panel <spreadsheet/pivot-tables/properties>`.
#. Click the :icon:`fa-cog` :guilabel:`(gear)` icon.
#. Click :icon:`fa-exchange` :guilabel:`Flip axes`.

  .. note::
     Depending on the volume of data, flipping the axes of a pivot table can result in a
     :guilabel:`#SPILL` error. This happens when a formula tries to output a range of values, but

>      something is preventing cells being populated, such as other data, merged cells, or the

Reviewer: ```suggestion
     something is preventing cells from being populated, such as other data, merged cells, or the
```

#. Right-click any value in the relevant column, then click :icon:`os-sort-range` :guilabel:`Sort
   pivot` and select :guilabel:`Ascending` or :guilabel:`Descending`.

To return to the default order, follow the same steps, then select :guilabel:`No sorting` from the
dropdown.

.. _spreadsheet/pivot-tables/dimensions-grouping:

Group dimension values
~~~~~~~~~~~~~~~~~~~~~~

Grouping dimension values allows you to declutter your pivot table by combining multiple dimension
values into a single, collapsible group. This can be useful, for example, for creating an
:guilabel:`Others` category to group less significant values and allow focus to be placed on more
significant data.

To group dimension values and create an :guilabel:`Others` category:

#. Open the :ref:`pivot table's properties panel <spreadsheet/pivot-tables/properties>`.
#. Hold `Ctrl` then select the cells of the row or column dimension values that should be grouped.
#. Right-click one of the selected cells, then click :icon:`fa-plus-square-o` :guilabel:`Group
   pivot dimensions`. A new dimension is added to the :guilabel:`Columns` or :guilabel:`Rows`

>    section, as relevant. Change the sorting to :guilabel:`Ascending` or :guilabel:`Descending`.gi

Reviewer: ```suggestion
   section, as relevant. Change the sorting to :guilabel:`Ascending` or :guilabel:`Descending`.
```

#. On the card of the new dimension:

   - change the :guilabel:`Order by` field from :guilabel:`Unsorted` to :guilabel:`Ascending` or
     :guilabel:`Descending`, as desired
   - click :icon:`fa-angle-right` :guilabel:`Groups` to expand the sub-section
   - optionally, rename :guilabel:`Group` by clicking it and entering a new name
   - click :guilabel:`+ "Others"` group to have the unselected dimension values added to a single

>      group that is then placed after the selected values

Reviewer: ```suggestion
   - Change the :guilabel:`Order by` field from :guilabel:`Unsorted` to :guilabel:`Ascending` or
     :guilabel:`Descending`, as desired.
   - Click :icon:`fa-angle-right` :guilabel:`Groups` to expand the sub-section.
   - Optionally, rename :guilabel:`Group` by clicking it and entering a new name.
   - Click the :guilabel:`+ "Others"` group to have the unselected dimension values added to a single
     group that is then placed after the selected values.
```

 .. _spreadsheet/pivot-tables/measures:

Measures
--------

The measures of a pivot table represent what is being measured, or analyzed, and are typically
fields such as :guilabel:`Total`, :guilabel:`Quantity ordered`, etc.

.. tip::
   If the desired measure did not exist in the original data source, create a :ref:`calculated
   measure <spreadsheet/pivot-tables/measures-calculated>`, e.g., to show the average revenue per
   order.

It is possible to:

- :ref:`add or remove measures <spreadsheet/pivot-tables/measures-add>`
- :ref:`rearrange measures <spreadsheet/pivot-tables/measures-rearrange>`
- :ref:`change how measures are calculated and shown <spreadsheet/pivot-tables/measures-calculated>`

.. _spreadsheet/pivot-tables/measures-add:

Add or remove measures
~~~~~~~~~~~~~~~~~~~~~~

To add a measure to a pivot table:

#. Open the :ref:`pivot table's properties panel <spreadsheet/pivot-tables/properties>`.
#. In the :guilabel:`Measures` section, click :guilabel:`Add`.
#. Select the appropriate measure or click :icon:`os-formula` :guilabel:`Add calculated measure` to
   create and add a custom :ref:`calculated measure
   <spreadsheet/pivot-tables/measures-calculated>`.

> #. Optionally, edit the name of the measure by clicking then editing the measure's name.

Reviewer: ```suggestion
#. Optionally, edit the name of the measure by clicking and then editing the measure's name.
```

the dimension's card. To temporarily hide a measure instead of removing it, click the :icon:`fa-eye`
:guilabel:`(Hide)` icon; to make a previously hidden measure visible, click the :icon:`fa-eye-slash`
:guilabel:`(Show)` icon.

.. tip::
   - To simply display a count of the dimensions, rather than a quantifiable measure, select
     :guilabel:`Count` as the measure.
   - In the pivot table, a measure's label is shown in a second header row by default; :ref:`hide
     this row <spreadsheet/pivot-tables/measures-hide-titles>` if desired.
   - The same measure can be added multiple times with different aggregations.
   - If multiple measures are added, they are shown in columns in the order in which they are added;
     :ref:`change the order <spreadsheet/pivot-tables/measures-rearrange>` if desired.

.. _spreadsheet/pivot-tables/measures-calculated:

Create calculated measures
~~~~~~~~~~~~~~~~~~~~~~~~~~

It is possible to add one or more calculated measures if the desired measure(s) did not exist in the
original data source. For example, a calculated measure could be added to show the average revenue
per order or the profit margin per product.

.. note::

>    The fields needed for the calculation of the new measure must already been added to the pivot

Reviewer: ```suggestion
   The fields needed for the calculation of the new measure must already have been added to the pivot
```

   table as dimensions or measures.

To add a calculated measure:

#. Open the :ref:`pivot table's properties panel <spreadsheet/pivot-tables/properties>`.
#. In the :guilabel:`Measures` section, click :guilabel:`Add`.
#. Below the scrollable list, click :icon:`os-formula` :guilabel:`Add calculated measure`.
#. Rename the calculated measure by clicking on the name and typing.
#. Click on the line starting with `=`, then enter the relevant formula.

   .. tip::
      :doc:`Functions <functions>` can be used in the formula of a calculated measure.

#. Choose how the measure should be aggregated by selecting a value from the dropdown.

.. example::

>    In the below example, the average revenue per order is added by dividing the sum of the sales

Reviewer: ```suggestion
   In the example below, the average revenue per order is calculated by dividing the sum of the sales
```

   by the number of orders.

   .. image:: pivot_tables/calculated-measure.png
      :alt: Formula for a calculated measure

.. _spreadsheet/pivot-tables/measures-rearrange:

Rearrange measures
~~~~~~~~~~~~~~~~~~

In the pivot table, measures are shown in columns in the order in which they were added.

To change the order in which measures are shown:

#. Open the :ref:`pivot table's properties panel <spreadsheet/pivot-tables/properties>`.
#. Drag the measure’s card to the desired position.

.. _spreadsheet/pivot-tables/measures-computation:

'Show measure as'
~~~~~~~~~~~~~~~~~

By default, measures are shown as the actual returned value, with no additional calculations
performed. However, it is also possible to show measures in relation to other data, e.g., as a
:guilabel:`% of grand total` or ranked from smallest to largest.

To change how a measure is shown:

#. Open the :ref:`pivot table's properties panel <spreadsheet/pivot-tables/properties>`.

> #. On the measure's card, click the :icon:`fa-cog` :guilabel:`Show values as` icon.

Reviewer: ```suggestion
#. On the measure's card, click the :icon:`fa-cog` :guilabel:`(gear)` icon.
```

Author: Instead, I've updated this to: 

```suggestion
#. On the measure's card, click :icon:`fa-cog` :guilabel:`(Show values as)`.
```

Since there is actually a label when you hover over it. Looks ok?

#. Select the desired option from the dropdown menu.
#. Click :guilabel:`Save`.

.. note::
   When changing how the measures are shown, the pivot table data updates in real time even if
   :ref:`Defer updates <spreadsheet/pivot-tables/properties-defer-updates>` is enabled.

.. _spreadsheet/pivot-tables/measures-hide-titles:

Remove measure titles
~~~~~~~~~~~~~~~~~~~~~

By default, measure titles are shown in a second header row in a pivot table.

To remove this header row:

#. Open the :ref:`pivot table's properties panel <spreadsheet/pivot-tables/properties>`.

> #. Double click the top left cell of the pivot table to be able to edit the formula.

Reviewer: ```suggestion
#. Double-click the top left cell of the pivot table to be able to edit the formula.
```

example, you can see the same data aggregated by different dimensions or use :doc:`global filters
<global_filters>` to offset the date and create pivot tables that compare the current period's data
with a previous period.

To duplicate a pivot table:

#. With the :ref:`pivot table properties <spreadsheet/pivot-tables/properties>` open, click
   the :icon:`fa-cog` :guilabel:`(gear)` icon then :icon:`fa-clone` :guilabel:`Duplicate`.

   The duplicated pivot table is automatically inserted into a new sheet in the spreadsheet, with
   the pivot table properties open in the right panel.

#. Edit the :guilabel:`Name` in the properties panel and the sheet tab if needed.

The new data source is assigned the next available :ref:`pivot table ID
<spreadsheet/pivot-tables/properties-id>`. For example, if no other pivot tables have been inserted
in the meantime, duplicating *Pivot #1* results in the creation of *Pivot #2*.

.. note::
   - Duplicating a pivot table by copying and pasting it or by duplicating the sheet does not create
     a new data source. Any changes made to the pivot table's properties would therefore impact any
     copies of the pivot table.

>    - When a pivot table is duplicated, the new pivot table is by default a :ref:`dynamic pivot table

Reviewer: ```suggestion
   - When a pivot table is duplicated, the new pivot table is, by default, a :ref:`dynamic pivot table
```

     <spreadsheet/insert-pivot-table/static-vs-dynamic>` whose dimensions and measures :ref:`can be
     manipulated <spreadsheet/pivot-tables/build-manipulate>`.

.. _spreadsheet/pivot-tables/delete:

Delete a pivot table
--------------------

To fully delete a pivot table and its data source from a spreadsheet, follow these steps in any
order:

- Delete the pivot table using your preferred means, e.g., via keyboard commands, spreadsheet
  menus, or by deleting the sheet.
- From the :ref:`properties panel <spreadsheet/pivot-tables/properties>` of the relevant pivot
  table, click the :icon:`fa-cog` :guilabel:`(gear)` icon then :icon:`fa-trash` :guilabel:`Delete`.
  This deletes the data source of the pivot table.

.. note::
   Deleting the pivot table first results in a warning message beside the corresponding data

>    source in the :guilabel:`Data` menu, while deleting the data source first results in pivot table

Reviewer: ```suggestion
   source in the :guilabel:`Data` menu, while deleting the data source first results in a pivot table
```
